play.pokemonshowdown.com/src/panel-chat.tsx

```diff
@@ -757,6 +791,105 @@ export class ChatTextEntry extends preact.Component<{
 		}
 		return false;
 	}
+	// Taken from the old client — just tweaked a bit to:
+	// - Work with <pre> instead of <textarea>
+	// - Fit the way `room` now handles user info
+	// TODO - add support for commands tabcomplete
+	handleTabComplete(reverse: boolean) {
+		// Don't tab complete at the start of the text box.
+		let idx = this.getSelection().start;
+		if (idx === 0) return false;
+		let users = (this.props.room as ChatRoom)?.users ||
```

Comment:
This is a sign that `this.props.room` is typed incorrectly. There's no need for an unsafe cast here.